extra_foam/pipeline/f_zmq.py

```diff
@@ -215,3 +220,189 @@ def __iter__(self):
+    endpoint : str
+        server socket you want to connect to (only support TCP socket).
+    sock : str, optional
+        socket type - supported: REQ, SUB.
+    ser : str, DEPRECATED
+        Serialization protocol to use to decode the incoming message (default
+        is msgpack) - supported: msgpack.
+    context : zmq.Context
+        To run the Client's sockets using a provided ZeroMQ context.
+    timeout : int
+        Timeout on :method:`next` (in seconds)
+        Data transfered at the EuXFEL for Mega-pixels detectors can be very
+        large. Setting a too small timeout might end in never getting data.
+        Some example of transfer timing for 1Mpix detector (AGIPD, LPD):
+            32 pulses per train (125 MB): ~0.1 s
+            128 pulses per train (500 MB): ~0.4 s
+            350 pulses per train (1.37 GB): ~1 s
+    Raises
+    ------
+    NotImplementedError
+        if socket type or serialization algorythm is not supported.
+    ZMQError
+        if provided endpoint is not valid.
+    """
+    def __init__(self, endpoint, timeout=None, context=None, connect=True):
+        # TODO
+        # [ ] ask for devices
+        # [ ] ask for keys/channels for device
+        # [ ] server as MDL?
+        #   [ ] handles several client connections
+        #   [ ] spawn bound device per channel connection (client select shared/copy)
+
+        self.dumps = msgpack.Packer(use_bin_type=True).pack
+        self.loads = partial(msgpack.loads, raw=False,
+                             max_bin_len=0x7fffffff)
+        self.monitored = set()
+
+        self.ctx = context or zmq.Context()
+        self.request = self.ctx.socket(zmq.PAIR)
+        self.request.SNDTIMEO = 5000
+        self.request.RCVTIMEO = 5000
+        self.request.setsockopt(zmq.LINGER, 0)
+        self.request.connect(endpoint)
+
+        self.data = None
+        self.timeout = timeout
+        self.connected = False
+        self._hb = None
+
+        if connect:
+            self.connect()
+
+    def connect(self):
+        if self.connected:
+            raise RuntimeError('Client is already connected')
+        # receive socket interfaces for slow and pipeline data
+        msg = self.ask({'request': 'hello', 'hostname': gethostname(),
+                        'username': getuser(), 'timestamp': time()})
+        self.data = self.ctx.socket(zmq.PULL)
+        self.data.set_hwm(50)
+        if self.timeout is not None:
+            self.data.RCVTIMEO = int(1000 * self.timeout)
+        self.data.connect(msg['data_addr'])
+        self.data.connect(msg['pipe_addr'])
+        self.ask({'request': 'hello', 'status': 'connected'})
+
+        self.connected = True
+        self._hb = Thread(target=self._heartbeat, daemon=True)
+        self._hb.start()  # start pinging server
+
+    def _heartbeat(self):
+        while self.connected:
+            sleep(10)
+            try:
+                self.ask({'request': 'ping'})
+            except zmq.error.ZMQError:
```

Comment:
@tmichela As discussed here https://git.xfel.eu/gitlab/karaboDevices/PipeToZeroMQ/merge_requests/51#note_168347, it might be good to capture the exception and make a clean terminal. The `ping` message is simply used to keep the server alive so the client user does not need to know whether the ping succeeded or not.

Reply:
That's right, although if the server does not reply, that probably means it's not working properly. I guess there is a better way to expose that to the client... But that probably helps in extra-foam if you see something like "server did not reply to ping, check what's going on there..."

Reply:
Of course, it helps if there is some error message. The problem is that the `ping` is running in a daemon thread. It could be achieved by returning a state of the client in the `next` method.